packages/doenetml-worker-rust/lib-doenetml-core/src/new_core/graph_based_core.rs

```diff
@@ -208,4 +526,191 @@ impl Core {
+
+        let children_virtual_node = self.structure_graph.get_nth_child(&parent_node, 0).unwrap();
+        self.structure_graph
+            .prepend_edge(&children_virtual_node, &child_node);
+    }
+
+    /// Add to `structure_graph` the relationships from component `component_idx` extending component `referent_idx`.
+    /// - The children of `referent` become the first children of `component` (i.e., become _virtual children_)
+    /// - Attributes of `referent` that match attributes of `component` become backup attributes of `component`,
+    ///   (i.e., become _virtual_attributes_)
+    /// - If the component types of `component` and `referent` are the same,
+    ///   then state props of `referent` are used for the state props of `component`
+    fn add_component_extending_structure(
+        &mut self,
+        component_idx: ComponentIdx,
+        referent_idx: ComponentIdx,
+    ) {
+        let component_node = GraphNode::Component(component_idx);
+        let referent_node = GraphNode::Component(referent_idx);
+
+        // The virtual node corresponding to `referent`'s children is the first child of `component`,
+        // so that `referent`'s children will be the first children of `component`.
+        let component_children_virtual_node = self
+            .structure_graph
+            .get_nth_child(&component_node, 0)
+            .unwrap();
+        let referent_children_virtual_node = self
+            .structure_graph
+            .get_nth_child(&referent_node, 0)
+            .unwrap();
+        self.structure_graph.prepend_edge(
+            &component_children_virtual_node,
+            &referent_children_virtual_node,
+        );
+
+        // Attributes from `referent` with the same name as an attribute from `component`
+        // are added as dependency of the attribute from `component`
+        // so that they will be used as a fallback if component doesn't have those attributes.
+
+        // TODO: we could check if attributes of `component` were specified and then omit the dependency.
+        let component_attributes_virtual_node = self
+            .structure_graph
+            .get_nth_child(&component_node, 1)
+            .unwrap();
+        let referent_attributes_virtual_node = self
+            .structure_graph
+            .get_nth_child(&referent_node, 1)
+            .unwrap();
+
+        let referent_attributes = self.components[referent_idx].get_attribute_names();
+
+        for (attr_idx, attr_name) in self.components[component_idx]
+            .get_attribute_names()
+            .into_iter()
+            .enumerate()
+        {
+            if let Some(ref_attr_idx) = referent_attributes.iter().position(|a| *a == attr_name) {
+                // found a matching attribute. Add a link to the attributes
+                let comp_attr = self
+                    .structure_graph
+                    .get_nth_child(&component_attributes_virtual_node, attr_idx)
+                    .unwrap();
```

Comment:
Did I not already add a `get_attr` function to `DirectedGraph`?

Reply:
yep